etcdserver: TestTriggerSnap panics

```
2016-04-20 18:18:07.895200 I | etcdserver: setting up the initial cluster version to 2.0
2016-04-20 18:18:07.896920 W | etcdserver: the running cluster version(2.1.0) is higher than the maximum cluster version(2.0.0) supported
2016-04-20 18:18:07.897052 W | etcdserver: the running cluster version(2.0.0) is lower than the minimal cluster version(2.1.0) supported
2016-04-20 18:18:07.902038 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:07.902139 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:07.971792 N | etcdserver: removed member 2 from cluster 0
2016-04-20 18:18:07.973856 N | etcdserver: removed member 1 from cluster 0
2016-04-20 18:18:07.974887 N | etcdserver: removed member 3 from cluster 0
2016-04-20 18:18:07.975522 N | etcdserver: removed member 4 from cluster 0
2016-04-20 18:18:07.975822 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:07.975891 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:07.977094 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:07.977158 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:07.978035 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:07.978097 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:07.978917 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:07.978980 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:08.002534 I | etcdserver: set snapshot count to default 10000
2016-04-20 18:18:08.002648 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:08.030223 I | etcdserver: saved snapshot at index 1
2016-04-20 18:18:08.030377 I | etcdserver: compacted raft log at 1
2016-04-20 18:18:08.095448 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: to_be_decided]
2016-04-20 18:18:08.099401 I | etcdserver: start to snapshot (applied: 11, lastsnap: 0)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x7761e1]

goroutine 466 [running]:
github.com/coreos/etcd/storage/backend.(*batchTx).commit(0xc8204a2200, 0x12c1b00)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/batch_tx.go:163 +0x121
github.com/coreos/etcd/storage/backend.(*batchTx).Commit(0xc8204a2200)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/batch_tx.go:140 +0x74
github.com/coreos/etcd/storage/backend.(*backend).ForceCommit(0xc82049c960)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:130 +0x46
github.com/coreos/etcd/storage.(*store).Commit(0xc8201b22c0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/kvstore.go:301 +0x5b
github.com/coreos/etcd/storage.(*watchableStore).Commit(0xc820194620)
	<autogenerated>:45 +0x5c
github.com/coreos/etcd/etcdserver.(*EtcdServer).snapshot.func1(0x7fa3c744fcf0, 0xc8204a2280, 0xc820224400, 0xb, 0xc820481170)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:1394 +0x522
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).snapshot
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:1422 +0x118

goroutine 1 [chan receive]:
testing.RunTests(0x12c3978, 0x160bbe0, 0x27, 0x27, 0xee6601)
	/tmp/workdir/go/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc820499f00, 0x1613780)
	/tmp/workdir/go/src/testing/testing.go:494 +0xe5
main.main()
	github.com/coreos/etcd/etcdserver/_test/_testmain.go:206 +0x385

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 108 [select]:
github.com/coreos/etcd/storage.(*watchableStore).syncWatchersLoop(0xc820194a10)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/watchable_store.go:237 +0x1a1
created by github.com/coreos/etcd/storage.newWatchableStore
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/watchable_store.go:77 +0x4bd

goroutine 410 [chan receive]:
github.com/coreos/etcd/etcdserver.(*EtcdServer).Stop(0xc820224400)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:883 +0x186
github.com/coreos/etcd/etcdserver.TestTriggerSnap(0xc820180120)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server_test.go:930 +0x121c
testing.tRunner(0xc820180120, 0x160be80)
	/tmp/workdir/go/src/testing/testing.go:456 +0xdd
created by testing.RunTests
	/tmp/workdir/go/src/testing/testing.go:561 +0xaa4

goroutine 106 [select]:
github.com/coreos/etcd/storage/backend.(*backend).run(0xc82020f980)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:182 +0x1d5
created by github.com/coreos/etcd/storage/backend.newBackend
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:117 +0x437

goroutine 45 [chan receive]:
github.com/coreos/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc82015f180)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:174 +0xa3
created by github.com/coreos/etcd/pkg/logutil.NewMergeLogger
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:92 +0x11e

goroutine 107 [select]:
github.com/coreos/etcd/pkg/schedule.(*fifo).run(0xc82020faa0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:146 +0x459
created by github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x366

goroutine 415 [select]:
github.com/coreos/etcd/pkg/testutil.(*recorderStream).Wait(0xc8200f2230, 0xc, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/testutil/recorder.go:113 +0x5ea
github.com/coreos/etcd/etcdserver.TestTriggerSnap.func2(0xa, 0xc8204a22c0, 0xc820180120, 0xc82049cea0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server_test.go:913 +0xb5
created by github.com/coreos/etcd/etcdserver.TestTriggerSnap
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server_test.go:924 +0x10dd

goroutine 414 [runnable]:
github.com/coreos/etcd/storage/backend.(*backend).Close(0xc82049c960, 0x0, 0x0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:194 +0x90
github.com/coreos/etcd/etcdserver.(*EtcdServer).run.func1(0xc820224400, 0x7fa3c7450458, 0xc82049d1a0)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:660 +0x25e
github.com/coreos/etcd/etcdserver.(*EtcdServer).run(0xc820224400)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:703 +0xb23
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).start
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:541 +0x6ab

goroutine 454 [select]:
github.com/coreos/etcd/storage/backend.(*backend).run(0xc8204203c0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:182 +0x1d5
created by github.com/coreos/etcd/storage/backend.newBackend
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:117 +0x437

goroutine 455 [select]:
github.com/coreos/etcd/pkg/schedule.(*fifo).run(0xc8204204e0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:146 +0x459
created by github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x366

goroutine 456 [select]:
github.com/coreos/etcd/storage.(*watchableStore).syncWatchersLoop(0xc8201c0150)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/watchable_store.go:237 +0x1a1
created by github.com/coreos/etcd/storage.newWatchableStore
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/storage/watchable_store.go:77 +0x4bd
FAIL	github.com/coreos/etcd/etcdserver	2.888s
```

possibly related to 74d50884bbdf64f943e818068c5632d66cda07fe

/cc @xiang90

--- a/etcdserver/server.go
+++ b/etcdserver/server.go
@@ -208,6 +208,10 @@ type EtcdServer struct {
 	forceVersionC chan struct{}
 
 	msgSnapC chan raftpb.Message
+
+	// wg is used to wait for the go routines that depends on the server state
+	// to exit when stopping the server.
+	wg sync.WaitGroup
 }
 
 // NewServer creates a new EtcdServer from the supplied configuration. The
@@ -535,6 +539,8 @@ func (s *EtcdServer) run() {
 	defer func() {
 		s.r.stop()
 		sched.Stop()
+
+		s.wg.Wait()
 
 		// kv, lessor and backend can be nil if running without v3 enabled
 		// or running unit tests.
@@ -1089,7 +1095,10 @@ func (s *EtcdServer) applyConfChange(cc raftpb.ConfChange, confState *raftpb.Con
 func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
 	clone := s.store.Clone()
 
+	s.wg.Add(1)
 	go func() {
+		defer s.wg.Done()
+
 		d, err := clone.SaveNoCopy()
 		// TODO: current store will never fail to do a snapshot
 		// what should we do if the store might fail?

--- a/storage/backend/backend.go
+++ b/storage/backend/backend.go
@@ -201,6 +201,18 @@ func (b *backend) Commits() int64 {
 }
 
 func (b *backend) Defrag() error {
+	err := b.defrag()
+	if err != nil {
+		return err
+	}
+
+	// commit to update metadata like db.size
+	b.batchTx.Commit()
+
+	return nil
+}
+
+func (b *backend) defrag() error {
 	// TODO: make this non-blocking?
 	// lock batchTx to ensure nobody is using previous tx, and then
 	// close previous ongoing tx.
@@ -251,8 +263,6 @@ func (b *backend) Defrag() error {
 	if err != nil {
 		log.Fatalf("backend: cannot begin tx (%s)", err)
 	}
-	// commit to update metadata like db.size
-	b.batchTx.commit(false)
 
 	return nil
 }

--- a/storage/backend/batch_tx.go
+++ b/storage/backend/batch_tx.go
@@ -160,6 +160,8 @@ func (t *batchTx) commit(stop bool) {
 	// commit the last tx
 	if t.tx != nil {
 		if t.pending == 0 && !stop {
+			t.backend.mu.RLock()
+			defer t.backend.mu.RUnlock()
 			atomic.StoreInt64(&t.backend.size, t.tx.Size())
 			return
 		}

